File: LeetCode75_Artur/Binary Tree - DFS/1372_Longest_ZigZag_Path_in_a_Binary_Tree.py
"""
https://leetcode.com/problems/longest-zigzag-path-in-a-binary-tree/description/?envType=study-plan-v2&envId=leetcode-75
"""

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

        
class Solution:
    def __init__(self) -> None:
        self.__longest_len = 0
    
    def dfs(self, root: Optional[TreeNode], is_left: bool, curr_len: int) -> None:
        if not root:
            return 
        
        if curr_len > self.__longest_len:
            self.__longest_len = curr_len
            
        if is_left:
            self.dfs(root.left, True, 1)
            self.dfs(root.right, False, curr_len + 1)
        else:
            self.dfs(root.left, True, curr_len + 1)
            self.dfs(root.right, False, 1)
    
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        self.dfs(root.left, True, 1)
        self.dfs(root.right, False, 1)
        
        return self.__longest_len
        
        
if __name__ == "__main__":
    tree1Node8 = TreeNode(1)
    tree1Node7 = TreeNode(1, None, tree1Node8)
    tree1Node6 = TreeNode(1)
    tree1Node5 = TreeNode(1, None, tree1Node7)
    tree1Node4 = TreeNode(1, tree1Node5, tree1Node6)
    tree1Node3 = TreeNode(1)
    tree1Node2 = TreeNode(1, tree1Node3, tree1Node4)
    tree1Node1 = TreeNode(1, None, tree1Node2)
    
    tree2Node7 = TreeNode(1)
    tree2Node6 = TreeNode(1)
    tree2Node5 = TreeNode(1, None, tree2Node7)
    tree2Node4 = TreeNode(1, tree2Node5, tree2Node6)
    tree2Node3 = TreeNode(1)
    tree2Node2 = TreeNode(1, None, tree2Node4)
    tree2Node1 = TreeNode(1, tree2Node2, tree2Node3)
    
    tree3Node1 = TreeNode(1)
    
    print(Solution().longestZigZag(tree1Node1))
    print(Solution().longestZigZag(tree2Node1))
    print(Solution().longestZigZag(tree3Node1))

File: LeetCode75_Artur/Binary Tree - DFS/1372_Longest_ZigZag_Path_in_a_Binary_Tree.cpp
/*
 https://leetcode.com/problems/longest-zigzag-path-in-a-binary-tree/description/?envType=study-plan-v2&envId=leetcode-75
*/

#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;


struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution
{
private:
    int longestWay;
public:
    void dfs(TreeNode* root, bool isLeft, int currLen)
    {
        if( !root )
            return;

        if( currLen > longestWay )
            longestWay = currLen;

        if( isLeft )
        {
            dfs(root->right, false, currLen + 1);
            dfs(root->left, true, 1);
        }
        else
        {
            dfs(root->left, true, currLen + 1);
            dfs(root->right, false, 1);
        }
    }

    int longestZigZag(TreeNode* root)
    {
        longestWay = 0;
        dfs(root->left, true, 1);
        dfs(root->right, false, 1);

        return longestWay;
    }
};


int main()
{
    Solution solution;

    auto* tree1Node8 = new TreeNode(1);
    auto* tree1Node7 = new TreeNode(1, nullptr, tree1Node8);
    auto* tree1Node6 = new TreeNode(1);
    auto* tree1Node5 = new TreeNode(1, nullptr, tree1Node7);
    auto* tree1Node4 = new TreeNode(1, tree1Node5, tree1Node6);
    auto* tree1Node3 = new TreeNode(1);
    auto* tree1Node2 = new TreeNode(1, tree1Node3, tree1Node4);
    auto* tree1Node1 = new TreeNode(1, nullptr, tree1Node2);

    auto* tree2Node7 = new TreeNode(1);
    auto* tree2Node6 = new TreeNode(1);
    auto* tree2Node5 = new TreeNode(1, nullptr, tree2Node7);
    auto* tree2Node4 = new TreeNode(1, tree2Node5, tree2Node6);
    auto* tree2Node3 = new TreeNode(1);
    auto* tree2Node2 = new TreeNode(1, nullptr, tree2Node4);
    auto* tree2Node1 = new TreeNode(1, tree2Node2, tree2Node3);

    auto* tree3Node1 = new TreeNode(1);

    cout << solution.longestZigZag(tree1Node1) << endl;
    cout << solution.longestZigZag(tree2Node1) << endl;
    cout << solution.longestZigZag(tree3Node1) << endl;

    return 0;
}

/*
 Main idea:
 Проходимось за допомогою dfs, який має два додаткових параметри:
 - isLeft — чи до цього вузла ми дійшли рухом на ліво;
 - currLen — поточна довжина зігзагу

 Якщо ми пішли до вузла наліво, то для dfs, який піде направо довжина
 зігзагу становитиме на 1 більше і значення isLeft false, для лівого
 сина ситуація буде протилежною. Ми збили зігзаг, тому поточний вузол
 буде першим у новому зігзагу, а його лівий син буде наступним вузлом,
 тому туди передається довжина 1. Для вузла, у якого значення isLeft
 false буде інша, для правого сина значення довжина 1, а для лівого
 поточна довжина +1.

 На кожному виклиці dfs перевіряється чи поточна довжина є більша за
 максимальну, якщо так, то це значення переписується.

 Часова складність алгоритму: O(n).
*/
